app/controlplane/internal/usercontext/apitoken_middleware.go

```diff
@@ -101,6 +102,98 @@ func WithCurrentAPITokenAndOrgMiddleware(apiTokenUC *biz.APITokenUseCase, orgUC
 	}
 }
 
+// WithCurrentAPITokenAndRobotAccountMiddleware injects the API-Token, organization + robot account to the context
+func WithCurrentAPITokenAndRobotAccountMiddleware(apiTokenUC *biz.APITokenUseCase, orgUC *biz.OrganizationUseCase, wfUC *biz.WorkflowUseCase, logger *log.Helper) middleware.Middleware {
+	return func(handler middleware.Handler) middleware.Handler {
+		return func(ctx context.Context, req interface{}) (interface{}, error) {
+			authInfo, ok := FromJWTAuthContext(ctx)
+			// If not found means that there is no currentUser set in the context
+			if !ok {
+				logger.Warn("couldn't extract org/user, JWT parser middleware not running before this one?")
+				return nil, errors.New("can't extract JWT info from the context")
+			}
+
+			// If the token is not an API token, we don't need to do anything
+			if !authInfo.IsAPIToken {
+				return handler(ctx, req)
+			}
+
+			genericClaims, ok := authInfo.Claims.(jwt.MapClaims)
+			if !ok {
+				return nil, errors.New("error mapping the claims")
+			}
+
+			// We've received an API-token
+			if genericClaims.VerifyAudience(apitoken.Audience, true) {
+				var err error
+				tokenID, ok := genericClaims["jti"].(string)
+				if !ok || tokenID == "" {
+					return nil, errors.New("error mapping the API-token claims")
+				}
+
+				ctx, err = setRobotAccountFromAPIToken(ctx, apiTokenUC, orgUC, wfUC, req, tokenID)
+				if err != nil {
+					return nil, fmt.Errorf("error setting current org and robot account: %w", err)
+				}
+
+				logger.Infow("msg", "[authN] processed credentials", "id", tokenID, "type", "API-token")
+			}
+
+			return handler(ctx, req)
+		}
+	}
+}
+
+func setRobotAccountFromAPIToken(ctx context.Context, apiTokenUC *biz.APITokenUseCase, orgUC *biz.OrganizationUseCase, wfUC *biz.WorkflowUseCase, req interface{}, tokenID string) (context.Context, error) {
+	if tokenID == "" {
+		return nil, errors.New("error retrieving the key ID from the API token")
+	}
+
+	// Check that the token exists and is not revoked
+	token, err := apiTokenUC.FindByID(ctx, tokenID)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving the API token: %w", err)
+	} else if token == nil {
+		return nil, errors.New("API token not found")
+	}
+
+	// Note: Expiration time does not need to be checked because that's done at the JWT
+	// verification layer, which happens before this middleware is called
+	if token.RevokedAt != nil {
+		return nil, errors.New("API token revoked")
+	}
+
+	// Find the associated organization
+	org, err := orgUC.FindByID(ctx, token.OrganizationID.String())
```

Comment:
do you need to load the org? Can't you just use the `token.OrganizationID.String()` note that at this point the token is authenticated already